### PR TITLE
feat: Backport RWDB 

### PR DESCRIPTION
**Pull request by shortthefomo**
## High Level Overview of Change

Port Xahau's **RWDB** in-memory node store to xrpld ([Xahau/xahaud#378](https://github.com/Xahau/xahaud/pull/378)). `[node_db] type=rwdb` is always null-backend: `fetch()` returns `NotFound`, `store()` is a no-op, and ledger state is kept alive through `Ledger` → SHAMap `shared_ptr` chains plus a capped sliding window of recent ledgers. A synced node uses about 11–13 GB RAM in testing.

Pair with `[relational_db] backend=rwdb` for the in-memory relational store (and in-memory peer-finder). Intended for validators, submission, and pathfinding nodes that do not need a local historical node store (Clio or another slave can persist).

### Context of Change

[XLS discussion](https://github.com/XRPLF/XRPL-Standards/discussions/559). Disk backends (NuDB/RocksDB) are unnecessary I/O and SSD cost for nodes that only need recent ledgers. There is no non-null RWDB path.

How it works:

- `type=rwdb` registers a null `RWDBBackend` and publishes a process-wide refcounted flag (`NullBackendFlag`) so libxrpl can detect null mode without an env var.
- Node store is a plain (non-rotating) `Database`. Rotation skips `clearCaches` / `makeBackendRotating` / `rotate`; only relational `clearPrior` runs. Evicting TreeNodeCache would be unrecoverable.
- `ledger_history` must be `> 0`. If `online_delete` is omitted it defaults to `max(ledger_history, minimum interval)`.
- Shared `FullBelowCache` is disabled in null mode. Inbound ledgers are primed with a delta walk from the closest fully-wired same-chain base, or by trusting sync pinning when no base exists.
- `retainedLedgers_` pins at most 256 recent ledgers, and only in null-backend mode.

### API Impact

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [x] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

## Config

[node_db] type=rwdb advisory_delete=0

[relational_db] backend=rwdb

`ledger_history` must be `> 0`. Set both sections; `type=rwdb` alone still creates SQLite relational files. No extra keys or environment variables.

### CODE REVIEW

**Comment by shortthefomo** [comment]
Putting this on hold have successfully got a null backend working so as to totally void this in place copy and rotation of the db as this makes no sense for a in memory backend. after testing will up date this branch and reopen PR.

**Comment by Tokeiito** [comment]
Love it. Would be great to have in PR information about use cases. Where this fits and reasoning behind it. Such information could be used while updating documentation and understanding PR itself.

**Comment by shortthefomo** [comment]
> PR information about use cases. Where this fits and reasoning behind it. Such information could be used while updating documentation and understanding PR itself.


leave some here ill add the above later.

1. memory nodes are lighter and faster. great for validators, submission nodes.
2. memory nodes can be paired to a slave system that write data to backup. eg Clio or other.
3. memory nodes are great for pathfinding nodes (other PRs by me opening potential of this node type up).

basically if you need todo anything with the XRPL and don't require history this is the node type that should perform the best.

**Comment by godexsoft** [comment]
We recently merged a refactor to `develop` that enables clang-tidy's `readability-identifier-naming`. Your branch now has heavy conflicts that are largely mechanical. Below is a workflow that aligns your branch's naming with develop _before_ merging, which should minimize the merge conflicts.

## One-time setup

If you don't already have clang-tidy working in your env, on macOS:

```bash
brew install llvm@21
# Follow brew's hint to put $(brew --prefix llvm@21)/bin on PATH so run-clang-tidy is found.
```

## Workflow on your branch (before merging develop)

**1. Grab the new `.clang-tidy` from develop without pulling anything else.** Sync your fork on GitHub first, then:

```bash
git remote -v   # should show 'upstream' among others; if not:
# git remote set-url upstream git@github.com:XRPLF/rippled.git
git fetch upstream
git checkout upstream/develop -- .clang-tidy
```

**2. Reconfigure conan/cmake so `compile_commands.json` is fresh.**

**3. Apply renames for the files modified in your PR:**

```bash
git diff --name-only $(git merge-base HEAD upstream/develop) HEAD \
  | grep -E '\.(cpp|h|hpp|ipp)$' \
  | xargs run-clang-tidy -p build -fix -allow-no-checks
# or -p .build, or whatever your build dir is called
```

**4. Build + test, then commit as a single dedicated commit:**

```bash
cmake --build build -j8
git commit -am "refactor: Align identifier naming with develop"
```

**5. Now merge develop:**

```bash
git merge upstream/develop
```

## Extra

Run clang-tidy once more after the merge to catch any stragglers introduced from develop's side:

```bash
run-clang-tidy -p build -fix -allow-no-checks src tests
# or -p .build, or whatever your build dir is called
```

File: include/xrpl/basics/ReaderPreferringSharedMutex.h [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    unlock()
+    {
+        pthread_rwlock_unlock(&rwlock_);
+    }
+
+    // Shared (reader) locking
+    void
+    lock_shared()
+    {
+        pthread_rwlock_rdlock(&rwlock_);
```
[the remark this answers is not in the document]
Comment (shortthefomo) [reply include/xrpl/basics/ReaderPreferringSharedMutex.h]: fixed in https://github.com/XRPLF/rippled/pull/6549/commits/43056eddab2cbdbe34b0af9a4e5dbad5a6c8e8dd

File: src/libxrpl/nodestore/backend/RWDBFactory.cpp [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+            old.swap(table_);  // O(1) swap; release lock before destructor runs
+        }
+        // 'old' is now destroyed outside the lock — no fetch() can be
+        // blocked by the (potentially millions-of-entries) map destructor.
+    }
+
+    static bool
+    nullMode()
+    {
+        char const* e = std::getenv("XRPL_RWDB_NULL");
```
[the remark this answers is not in the document]
Comment (shortthefomo) [reply src/libxrpl/nodestore/backend/RWDBFactory.cpp]: fixed in https://github.com/XRPLF/rippled/pull/6549/commits/43056eddab2cbdbe34b0af9a4e5dbad5a6c8e8dd

**Comment by sublimator** [comment]
@shortthefomo 

re: nullfactory, this might be of interest, AI dump, so ... dump to AI, haha 

https://github.com/XRPLF/rippled/issues/7870


(edit: direct gist link): https://gist.github.com/sublimator/d89356a3933f6c64432089810440afeb 

Been a while since I had time to mess around with shamap fun, but that was an annoying source of the missing nodes you might see with this now and then

**Comment by mvadari** [comment]
<details>
<summary>I had Claude do a review of this PR (collapsed here because it's fairly lengthy). It had a lot of serious concerns.</summary>

## Overview

Ports Xahau's **RWDB** in-memory node store to xrpld. Setting `type=rwdb` in `[node_db]` flips a process-wide "null backend" flag: `fetch()` always returns `NotFound`, `store()` is a no-op, and ledger state is kept alive purely through `Ledger → SHAMap` `shared_ptr` chains plus a sliding window of pinned recent ledgers. Also adds `RWDBDatabase` (in-memory `RelationalDatabase`), `InMemoryStore` (peer-finder), a reader-preferring mutex, and "fully wired" tracking on `Ledger`.

The concept is reasonable and the use cases (validators, submission/pathfinding nodes) are real. But as submitted this PR carries a lot of unrelated, high-risk collateral and several correctness problems that affect **non-RWDB** operators. It's not close to mergeable in its current shape.

---

## Blocking issues

### 1. Build-system changes are local workarounds that break everyone else
[[cmake/deps/Boost.cmake](https://claude.ai/epitaxy/cmake/deps/Boost.cmake:4)](cmake/deps/Boost.cmake:4) deletes the entire `find_package(Boost REQUIRED COMPONENTS ...)` call while [[line 23](https://claude.ai/epitaxy/cmake/deps/Boost.cmake:23)](cmake/deps/Boost.cmake:23) still links `Boost::chrono`, `Boost::container`, `Boost::context`, `Boost::json`, etc. Those imported targets no longer exist, and the version/`REQUIRED` guard is gone. [[conan/profiles/default](https://claude.ai/epitaxy/conan/profiles/default:55)](conan/profiles/default:55) gains `boost*:shared=False` — a global change to every developer's and CI's dependency graph. Both are commented as fixes for one macOS `boost_unit_test_framework` problem. These must be dropped from the PR.

### 2. `retainedLedgers_` pins ledgers on every node type, unbounded with `ledger_history=full`
[[LedgerMaster.cpp:834](https://claude.ai/epitaxy/src/xrpld/app/ledger/detail/LedgerMaster.cpp:834)](src/xrpld/app/ledger/detail/LedgerMaster.cpp:834) gates only on `isCurrent && ledgerHistorySize_ > 0` — not on null mode. `ledger_history=full` sets `ledgerHistory = UINT32_MAX` ([[Config.cpp:691](https://claude.ai/epitaxy/src/xrpld/core/detail/Config.cpp:691)](src/xrpld/core/detail/Config.cpp:691)), so a full-history NuDB node will hard-pin every validated ledger and its SHAMap in RAM forever. Even at the default 256 this is a hard, non-evictable retention on top of the existing `LedgerHistory` TaggedCache. Gate on `isNullBackend()` and clamp the window independently of `ledger_history`.

### 3. `logThrow()` now captures and prints a stacktrace on every throw
[[contract.cpp:19](https://claude.ai/epitaxy/src/libxrpl/basics/contract.cpp:19)](src/libxrpl/basics/contract.cpp:19) builds a `boost::stacktrace::stacktrace()` and writes it to both the log and `std::cerr` unconditionally. `Throw<>` is used throughout the codebase, including non-exceptional paths; stack capture plus symbolization is orders of magnitude more expensive than the previous single log line, and the unconditional `std::cerr` write is unfiltered by log level. This is debugging scaffolding — it should not ship, or at minimum be gated behind null mode/a debug flag. (Also note nothing links a `boost_stacktrace_*` backend, so output quality is platform-dependent.)

### 4. Consensus heartbeat rework is unrelated and risky
[[NetworkOPs.cpp:1068](https://claude.ai/epitaxy/src/xrpld/app/misc/NetworkOPs.cpp:1068)](src/xrpld/app/misc/NetworkOPs.cpp:1068) moves `processHeartbeatTimer()` off the JobQueue onto the io_service thread, and [[line 1113](https://claude.ai/epitaxy/src/xrpld/app/misc/NetworkOPs.cpp:1113)](src/xrpld/app/misc/NetworkOPs.cpp:1113) switches `masterMutex` acquisition to `try_to_lock`, skipping peer-count/mode transitions on contention. That means `consensus_.timerEntry()` now runs on an io worker and `LoadManager::heartbeat()` can be skipped for arbitrarily long — the load manager's stall detection depends on it. This is a substantive consensus-scheduling change that deserves its own PR with its own justification and measurements.

### 5. Process-wide mutable global for null mode
[[NullBackendFlag.h](https://claude.ai/epitaxy/include/xrpl/basics/NullBackendFlag.h:17)](include/xrpl/basics/NullBackendFlag.h:17) is set in the `SHAMapStoreImp` constructor and cleared in its destructor with no refcount. rippled's unit test runner executes suites concurrently in one process; a single RWDB test suite flips global behavior (`SHAMapSync`'s FullBelowCache, `Ledger::fullWireForUse`, `primeInboundLedgerForUse`) for every other suite running at the same time, and the destructor clears it even if another `SHAMapStoreImp` still needs it. This will produce flaky, hard-to-diagnose failures. Prefer threading the flag through `Application`/`Config` (which is already available at nearly every call site) or, if a global is unavoidable, refcount it.

Relatedly, [[Config::nullBackend()](https://claude.ai/epitaxy/src/xrpld/core/Config.h:351)](src/xrpld/core/Config.h:351) is a `static` member that just forwards to the free function and reads no config state — two spellings of the same thing, and misleading on a `Config` object. Pick one.

### 6. `getTransaction` range scan is O(range) under a shared lock
[[RWDBDatabase.h:3660](https://claude.ai/epitaxy/src/xrpld/app/rdb/backend/RWDBDatabase.h)](src/xrpld/app/rdb/backend/RWDBDatabase.h) fixed the inverted-range overflow, but still walks `for (LedgerIndex seq = range->first(); ; ++seq)` doing a `ledgers_.find()` per sequence. A client-supplied wide range means millions of map lookups while holding `mutex_` shared, blocking writers. Count with `lower_bound`/`upper_bound` iteration over the actual entries instead — it's O(present entries) and trivially correct.

---

## Correctness / design concerns

- **`RWDBBackend` is entirely dead code.** Since `type=rwdb` unconditionally sets null mode, and `RWDBBackend` is only constructed for `type=rwdb`, `table_`, `store()`, `fetch()`, and `forEach()` can never do anything. A `registerNullFactory` already exists. Either delete the map and make it an explicit null backend, or justify why a second null implementation is needed.
- **`RWDBBackend::fetchBatch` doesn't override anything** — `node_store::Backend` has no `fetchBatch` ([[Backend.h](https://claude.ai/epitaxy/include/xrpl/nodestore/Backend.h:101)](include/xrpl/nodestore/Backend.h:101)). It's unreachable. Same for the unused `DecodedBlob.h`/`EncodedBlob.h`/`codec.h` includes.
- **Three near-identical copies of the same-chain-distance search**: [[InboundLedger.cpp](https://claude.ai/epitaxy/src/xrpld/app/ledger/detail/InboundLedger.cpp)](src/xrpld/app/ledger/detail/InboundLedger.cpp) (`sameChainDistance`), [[InboundLedgersImp::getClosestFullyWiredLedger](https://claude.ai/epitaxy/src/xrpld/app/ledger/detail/InboundLedgers.cpp)](src/xrpld/app/ledger/detail/InboundLedgers.cpp), and [[LedgerMaster::getClosestFullyWiredLedger](https://claude.ai/epitaxy/src/xrpld/app/ledger/detail/LedgerMaster.cpp)](src/xrpld/app/ledger/detail/LedgerMaster.cpp). ~40 lines each, byte-for-byte equivalent logic. Factor into one helper.
- **`wireCompleteSHAMap` is duplicated** in [[Ledger.cpp:52](https://claude.ai/epitaxy/src/libxrpl/ledger/Ledger.cpp)](src/libxrpl/ledger/Ledger.cpp) and [[InboundLedger.cpp:63](https://claude.ai/epitaxy/src/xrpld/app/ledger/detail/InboundLedger.cpp)](src/xrpld/app/ledger/detail/InboundLedger.cpp), and the `(void)item; ++leaves;` loop is an obscure way to say "force materialization" — a comment or a named helper would help a lot.
- **`Ledger(prevLedger, closeTime)` inherits `fullyWired_` from the parent.** The child has a different (mutated) state map; inheriting "fully wired" asserts something about maps that weren't walked. It may be true in practice, but the reasoning isn't recorded anywhere.
- **`setFullyWired() const` mutating a `mutable` atomic** on an otherwise-immutable object is surprising API. Worth a doc comment on why const is required here.
- **`inboundLedgerJobLimit` jumps 5 → 500** for RWDB ([[InboundLedger.cpp](https://claude.ai/epitaxy/src/xrpld/app/ledger/detail/InboundLedger.cpp)](src/xrpld/app/ledger/detail/InboundLedger.cpp)). That's a 100× increase in concurrent `JtLedgerData` jobs with no stated basis; it can starve other job types.
- **RWDB account-tx markers use a synthetic `txnSeq`** (vector index within the per-account bucket) rather than the ledger's real `TxnSeq`, so markers are not interchangeable with SQLite-produced markers. Fine internally, but it's an undocumented behavioral divergence in a public API surface.
- **`deleteTransactionByLedgerSeq`** clears `ledgers_[seq].transactions` and `transactionMap_` but leaves `accountTxMap_` entries in place, so `getAccountTransactionCount()` and `account_tx` disagree with `getTransactionCount()` afterwards.
- **`LedgerMaster.h` adds a `std::deque` member without `#include <deque>`** — the header currently has no such include.
- **`PeerImp.cpp`'s only change is an added include** (`SHAMapTreeNode.h`) with no corresponding code — leftover.

---

## Tests

Coverage is thin relative to a 3,400-line change, and several new suites don't test the code they name:

- [[RWDBDatabase_test.cpp](https://claude.ai/epitaxy/src/test/app/RWDBDatabase_test.cpp)](src/test/app/RWDBDatabase_test.cpp) is a single `static_assert(is_base_of_v<...>)` — a compile-time tautology, zero behavioral coverage of a 1,060-line class.
- [[SHAMapSyncFullBelowCache_test.cpp](https://claude.ai/epitaxy/src/tests/libxrpl/shamap/SHAMapSyncFullBelowCache_test.cpp)](src/tests/libxrpl/shamap/SHAMapSyncFullBelowCache_test.cpp) computes `bool useFullBelowCache = !nullMode;` **in the test itself** and asserts on that local. It never touches `SHAMapSync.cpp`. It would pass if the production code were deleted.
- [[InboundLedger_test.cpp](https://claude.ai/epitaxy/src/test/app/InboundLedger_test.cpp)](src/test/app/InboundLedger_test.cpp) only asserts that an empty cache returns empty, and its own comment says the real path "is tested indirectly."
- [[Config_test testRWDBOnlineDelete](https://claude.ai/epitaxy/src/test/core/Config_test.cpp)](src/test/core/Config_test.cpp) — all four cases are "should not throw," which is the default outcome; no assertion on the resulting configuration values.
- The RWDB tests in `AccountTx_test` and `Transaction_test` set `[relational_db] backend=rwdb` but leave `node_db` as `memory`, so the actual null-backend node-store path — the heart of this PR — is never exercised by any test.

What's missing and matters most: no test of the null-backend `fetch`-returns-`NotFound` path end to end, no test that a node in null mode can serve ledger data / survive rotation, no test of `primeInboundLedgerForUse` delta-walk vs. full-walk, and no test of `retainedLedgers_` bounding.

`ReaderPreferringSharedMutex_test` and `InMemoryStore_test` are genuinely good; note that `testExclusiveLock`'s `BEAST_EXPECT(!writerHoldsLock)` inside the reader lambda is timing-dependent and could flake.

---

## Smaller notes

- The PR description is substantially out of sync with the diff: it says `Config::null_backend()` was **removed** (it's added), describes an `XRPL_RWDB_NULL` env var (replaced by the atomic), lists `DatabaseRotatingImp` mutex changes and `PeerImp` TreeNodeCache fallback (neither is in the diff), and lists `PeerfinderManager`/`Application` RWDB wiring that isn't there. Please refresh it before review.
- `reader_preferring_shared_mutex` is only used by the dead `RWDBBackend`. Its constructor ignores the return of `pthread_rwlockattr_init`, and `unlock()`/`unlock_shared()` silently ignore errors.
- `RWDBDatabase.h` is 1,060 lines of fully-inline implementation in a header, contrary to the surrounding `src/xrpld/app/rdb/backend/detail/*.cpp` convention.
- `std::map<uint256 const, ...>` in `RWDBBackend` — the `const` on the key type is redundant and non-idiomatic.
- Several new files exceed the 100-column `ColumnLimit`; run `clang-format`.

---

## Recommendation

Split this. The RWDB feature itself deserves a focused PR; the Boost/Conan changes, the `logThrow` stacktrace, and the heartbeat/masterMutex rework are three separate, independently-reviewable changes that currently ride along unannounced. Before the RWDB part can land: fix the unconditional `retainedLedgers_` retention, resolve the global-flag/test-isolation problem, remove the dead backend code, deduplicate the three copies of the base-ledger search, and add tests that actually run a node with `type=rwdb`.

</details>

File: src/libxrpl/ledger/Ledger.cpp [hunk 3]
```diff
@@ -328,6 +334,29 @@ Ledger::setImmutable(bool rehash)
...
 
+bool
+Ledger::fullWireForUse(beast::Journal journal, char const* context) const
+{
+    if (!isNullBackend() || isFullyWired())
+        return true;
+
+    try
+    {
+        auto const stateLeaves = materializeSHAMapLeaves(stateMap_);
```
Comment (sublimator) [inline src/libxrpl/ledger/Ledger.cpp]: This is walking the full map each time ?
I think this was derived from that initial prototype hack to force linking and prove it was possible?
I can't recall off hand but I think there were more precise methods in the end

Comment (sublimator) [reply src/libxrpl/ledger/Ledger.cpp]: The document I linked before describes one of the annoying issues that might have been inspiring the hack. There's races for canonical inner nodes, and they throw away linked children instead of merging them, so you get the missing inner node

Comment (shortthefomo) [reply src/libxrpl/ledger/Ledger.cpp]: hehehe you catching me mid stream to night :) keep the comments coming appreciate it. (yip addressing as I go)

Comment (shortthefomo) [reply src/libxrpl/ledger/Ledger.cpp]: [45c5bb3](https://github.com/XRPLF/rippled/pull/6549/commits/45c5bb38d68be60f32092ec98c26fc1b38101485)
and
[21da542](https://github.com/XRPLF/rippled/pull/6549/commits/21da54251795d9b378d9fc9c7bb8af4157db6320) should address
